network/simple-onchain-discovery/src/lib.rs

```diff
@@ -185,78 +280,115 @@ impl ValidatorSetChangeListener {
+        );
 
         inc_by_with_context(
             &DISCOVERY_COUNTS,
             &self.network_context,
             "new_nodes",
-            updates.len() as u64,
-        );
-        info!(
-            NetworkSchema::new(&self.network_context),
-            "Update {} Network about new Node IDs",
-            self.network_context.network_id()
+            peer_set.len() as u64,
         );
 
-        for update in updates {
-            match self.conn_mgr_reqs_tx.send(update).await {
-                Ok(()) => (),
-                Err(e) => {
-                    inc_by_with_context(
-                        &DISCOVERY_COUNTS,
-                        &self.network_context,
-                        "send_failure",
-                        1,
-                    );
-                    warn!(
-                        NetworkSchema::new(&self.network_context),
-                        "Failed to send update to ConnectivityManager {}", e
-                    )
-                }
-            }
+        peer_set
+    }
+}
+
+/// Watcher for a file containing discovered peers at a given interval
+pub struct FileChangeListener {
+    network_context: Arc<NetworkContext>,
+    file_path: PathBuf,
+    update_channel: Mutex<channel::Sender<ConnectivityRequest>>,
+    time_service: TimeService,
+    interval_duration: Duration,
+    delay: Option<Pin<Box<Sleep>>>,
+}
+
+impl FileChangeListener {
+    pub fn new(
+        network_context: Arc<NetworkContext>,
+        file_path: &Path,
+        update_channel: channel::Sender<ConnectivityRequest>,
+        interval_duration: Duration,
+        time_service: TimeService,
+    ) -> Self {
+        FileChangeListener {
+            network_context,
+            file_path: file_path.to_path_buf(),
+            update_channel: Mutex::new(update_channel),
+            time_service,
+            interval_duration,
+            delay: None,
         }
     }
+}
 
-    /// Starts the listener to wait on reconfiguration events.
-    pub async fn start(mut self) {
-        info!(
-            NetworkSchema::new(&self.network_context),
-            "{} Starting OnChain Discovery actor", self.network_context
-        );
+#[async_trait]
+impl DiscoveryChangeListenerTrait for FileChangeListener {
+    fn network_context(&self) -> &NetworkContext {
+        &self.network_context
+    }
+
+    fn discovery_source(&self) -> DiscoverySource {
+        DiscoverySource::File
+    }
 
-        while let Some(payload) = self.next_reconfig_event().await {
-            self.process_payload(payload).await;
+    fn update_channel(&self) -> &Mutex<channel::Sender<ConnectivityRequest>> {
+        &self.update_channel
+    }
+}
+
+impl<'a> Stream for FileChangeListener {
+    type Item = PeerSet;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Wait for delay, or add the delay for next call
+        if let Some(ref mut delay) = self.delay {
```

Comment:
yuck, I guess we don't have access to unix events?

Reply:
That was sort of my thought, we should be able to leverage inotify or something to get a notification when the file has updated instead of having to reread it every X seconds

Reply:
could do something with the [`notify`](https://docs.rs/notify/5.0.0-pre.10/notify/index.html) crate

```rust
// in discovery init

let watcher = notify::immediate_watcher(|event| {
    let info = load_file("/path/to/file");
    let msg = ConnectivityRequest(DiscoverySource::File, info.into_peer_set());
    let _ = conn_mgr_reqs_tx.try_send(msg);
})?;
watcher.watch("/path/to/file", RecursiveMode::NonRecursive)?;

// store watcher
```

Reply:
There is a library to access unix file events that I found, but given our very tight controls on dependencies, I wanted to move fast and not argue with the team on whether I should pull in a library for what is mostly a stepping stone to on chain.

Reply:
@phlip9 that's the `crate` I mentioned, that I was skipping pulling in :P